core/src/main/java/org/bitcoinj/quorums/InstantSendManager.java

```diff
@@ -404,19 +407,46 @@ boolean processPendingInstantSendLocks() throws BlockStoreException {
     }
 
     boolean processPendingInstantSendLocks(boolean isDeterministic) throws BlockStoreException {
-        HashMap<Sha256Hash, Pair<Long, InstantSendLock>> pend;
+        HashMap<Sha256Hash, Pair<Long, InstantSendLock>> pend = new HashMap<>();
+        boolean moreWork = false;
 
-        lock.lock();
+        if (!isInstantSendEnabled()) {
+            return false;
+        }
 
+        lock.lock();
         try {
-            if (pendingInstantSendLocks.isEmpty())
-                return false;
-            pend = new HashMap<Sha256Hash, Pair<Long, InstantSendLock>>(pendingInstantSendLocks);
-            pendingInstantSendLocks = new HashMap<Sha256Hash, Pair<Long, InstantSendLock>>();
+            // only process a max 32 locks at a time to avoid duplicate verification of recovered signatures which have been
+            // verified by CSigningManager in parallel
+            final int maxCount = 32;
+            // The keys of the removed values are temporaily stored here to avoid invalidating an iterator
+            ArrayList<Sha256Hash> removed = Lists.newArrayListWithExpectedSize(maxCount);
+
+            //if (pendingInstantSendLocks.isEmpty())
+            //    return false;
+            for (Map.Entry<Sha256Hash, Pair<Long, InstantSendLock>> entry : pendingInstantSendLocks.entrySet()) {
+                Sha256Hash islockHash = entry.getKey();
+                InstantSendLock islock = entry.getValue().getSecond();
+
+                if (pend.size() >= maxCount) {
+                    moreWork = true;
+                    break;
+                }
+                // Check if we care about this islock on this run
+                if (islock.isDeterministic() != isDeterministic) {
+                    continue;
+                }
+                pend.put(islockHash, entry.getValue());
+                removed.add(islockHash);
+            }
+
+            for (Sha256Hash islockHash : removed) {
+                pendingInstantSendLocks.remove(islockHash);
+            }
 
             //try to process the invalidInstantSendLocks again
-            for (InstantSendLock isLock : invalidInstantSendLocks.keySet())
-                pendingInstantSendLocks.put(isLock.getHash(), new Pair<>(Long.valueOf(0L), isLock));
+            //for (InstantSendLock isLock : invalidInstantSendLocks.keySet())
+            //    pendingInstantSendLocks.put(isLock.getHash(), new Pair<>(Long.valueOf(0L), isLock));
```

Comment:
Not needed?

Reply:
Good Catch, I will put a TODO on this.  I am not sure.  In the past, this was added to retry ISLocks that had failed, but I am not sure if it ever worked or if it is still needed.

Reply:
One change with Core 0.17 is that when the Blockchain is synchronized, ISLocks are sent for old transactions after we get those transactions.  

This causes a problem, because the masternode lists and quorums that we keep track of are always up to date as of the current time (unless syncing is not complete, then the quorums are out to date).  Old transactions and their islocks will require old masternode lists and quorums to verify, but we don't keep track of those.

What was happening before this was commented out was that we would receive old ISLocks that couldn't be verified, so they would stay in the invalidInstantSendLocks map and be rechecked over and over and over again, and never be verified.